mailet/standard/src/main/java/org/apache/james/transport/mailets/ContactExtractor.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.mailet.Mailet;
+import org.apache.mailet.MailetException;
+import org.apache.mailet.base.GenericMailet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * <p>Collects the sender and the recipients of a message and store them as JSON in a specified message attribute.</p>
+ * <p>Here is the JSON format:</p>
+ * <pre><code>
+ * {
+ *   "userEmail" : "sender@james.org", 
+ *   "emails" : [ "to@james.org", "cc@james.org" ]
+ * }
+ * </code></pre>
+ * 
+ * <p>Sample configuration:</p>
+ *
+ * <pre><code>
+ * &lt;mailet match="All" class="ContactExtractor"&gt;
+ *   &lt;attribute&gt;ExtractedContacts&lt;/attribute&gt;
+ * &lt;/mailet&gt;
+ * </code></pre>
+ */
+public class ContactExtractor extends GenericMailet implements Mailet {
+
+    public static interface Configuration {
+        public static final String ATTRIBUTE = "attribute";
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ContactExtractor.class);
+
+    @VisibleForTesting ObjectMapper objectMapper;
+    private String extractAttributeTo;
+
+    @Override
+    public void init() throws MessagingException {
+        extractAttributeTo = getInitParameter(Configuration.ATTRIBUTE);
+        
+        if (Strings.isNullOrEmpty(extractAttributeTo)) {
+            throw new MailetException("No value for " + Configuration.ATTRIBUTE + " parameter was provided.");
+        }
+        objectMapper = new ObjectMapper().registerModule(new Jdk8Module());
+    }
+
+    @Override
+    public String getMailetInfo() {
+        return "ContactExtractor Mailet" ;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        try {
+            LOGGER.debug("extracting recipients");
+            Optional<String> payload = buildMessage(mail);
+            LOGGER.debug("payload : {}", payload);
+            payload.ifPresent(x -> mail.setAttribute(extractAttributeTo, x));
+        } catch (Exception e) {
+            LOGGER.error("Error while extracting contacts", e);
+        }
+    }
+
+    private Optional<String> buildMessage(Mail mail) {
+        return Optional.ofNullable(mail.getSender())
+                .map(MailAddress::asString)
+                .map(sender -> new Message(sender, recipients(mail)))
+                .map(Throwing.function(message -> objectMapper.writeValueAsString(message)));
+    }
+
+    private ImmutableList<String> recipients(Mail mail) {
+        return mail.getRecipients().stream()
+                .map(MailAddress::asString)
+                .collect(Guavate.toImmutableList());
+    }
+
+    public static class Message {
```

Comment:
IMO this "Message is a bad name. I would  prefer **Enveloppe**.

Reply:
Renamed